File: ml-agents-envs/mlagents_envs/side_channel/raw_bytes_channel.py
from mlagents_envs.side_channel import SideChannel, IncomingMessage, OutgoingMessage
from typing import List
import uuid


class RawBytesChannel(SideChannel):
    """
    This is an example of what the SideChannel for raw bytes exchange would
    look like. Is meant to be used for general research purpose.
    """

    def __init__(self, channel_id: uuid.UUID):
        self._received_messages: List[bytes] = []
        super().__init__(channel_id)

    def on_message_received(self, msg: IncomingMessage) -> None:
        """
        Is called by the environment to the side channel. Can be called
        multiple times per step if multiple messages are meant for that
        SideChannel.
        """
        self._received_messages.append(msg.get_raw_bytes())

    def get_and_clear_received_messages(self) -> List[bytes]:
        """
        returns a list of bytearray received from the environment.
        """
        result = list(self._received_messages)
        self._received_messages = []
        return result

    def send_raw_data(self, data: bytearray) -> None:
        """
        Queues a message to be sent by the environment at the next call to
        step.
        """
        msg = OutgoingMessage()
        msg.set_raw_bytes(data)
        super().queue_message_to_send(msg)

File: com.unity.ml-agents/Runtime/SideChannels/RawBytesChannel.cs
using System.Collections.Generic;
using System;

namespace Unity.MLAgents.SideChannels
{
    /// <summary>
    /// Side channel for managing raw bytes of data. It is up to the clients of this side channel
    /// to interpret the messages.
    /// </summary>
    public class RawBytesChannel : SideChannel
    {
        List<byte[]> m_MessagesReceived = new List<byte[]>();

        /// <summary>
        /// RawBytesChannel provides a way to exchange raw byte arrays between Unity and Python.
        /// </summary>
        /// <param name="channelId"> The identifier for the RawBytesChannel. Must be
        /// the same on Python and Unity.</param>
        public RawBytesChannel(Guid channelId)
        {
            ChannelId = channelId;
        }

        /// <inheritdoc/>
        protected override void OnMessageReceived(IncomingMessage msg)
        {
            m_MessagesReceived.Add(msg.GetRawBytes());
        }

        /// <summary>
        /// Sends the byte array message to the Python side channel. The message will be sent
        /// alongside the simulation step.
        /// </summary>
        /// <param name="data"> The byte array of data to send to Python.</param>
        public void SendRawBytes(byte[] data)
        {
            using (var msg = new OutgoingMessage())
            {
                msg.SetRawBytes(data);
                QueueMessageToSend(msg);
            }
        }

        /// <summary>
        /// Gets the messages that were sent by python since the last call to
        /// GetAndClearReceivedMessages.
        /// </summary>
        /// <returns> a list of byte array messages that Python has sent.</returns>
        public IList<byte[]> GetAndClearReceivedMessages()
        {
            var result = new List<byte[]>();
            result.AddRange(m_MessagesReceived);
            m_MessagesReceived.Clear();
            return result;
        }

        /// <summary>
        /// Gets the messages that were sent by python since the last call to
        /// GetAndClearReceivedMessages. Note that the messages received will not
        /// be cleared with a call to GetReceivedMessages.
        /// </summary>
        /// <returns> a list of byte array messages that Python has sent.</returns>
        public IList<byte[]> GetReceivedMessages()
        {
            var result = new List<byte[]>();
            result.AddRange(m_MessagesReceived);
            return result;
        }
    }
}
